### PR TITLE
CLOUDSTACK-10377: Fix Network restart for Nuage

### PR DESCRIPTION
**Pull request by fmaximus**
## Description

Changes in PR #2508 have caused network restart to fail in a Nuage setup,
as the new VR takes the same IP as the old one, and the old VR is still running.
Nuage doesn't support multiple VM's having the same IP.
We delay provisioning the interfaces in VSD until the old VR interface is released.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [x] I have added tests to cover my changes.
- [x] All relevant new and existing integration tests have passed.
- [x] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

File: plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java [hunk 1]
```diff
@@ -528,29 +533,35 @@ public void reserve(NicProfile nic, Network network, VirtualMachineProfile vm, D
...
-            if (answer == null || !answer.getResult()) {
-                s_logger.error("ReserveVmInterfaceNuageVspCommand failed for NIC " + nic.getId() + " attached to VM " + vm.getId() + " in network " + network.getId());
-                if ((null != answer) && (null != answer.getDetails())) {
-                    s_logger.error(answer.getDetails());
+                boolean defaultHasDns = getDefaultHasDns(networkHasDnsCache, nicFromDb);
+                VspDhcpVMOption dhcpOption = _nuageVspEntityBuilder.buildVmDhcpOption(nicFromDb, defaultHasDns, networkHasDns);
+                ReserveVmInterfaceVspCommand cmd = new ReserveVmInterfaceVspCommand(vspNetwork, vspVm, vspNic, vspStaticNat, dhcpOption);
+                Answer answer = _agentMgr.easySend(nuageVspHost.getId(), cmd);
+
+                if (answer == null || !answer.getResult()) {
```
Comment (rafaelweingartner) [inline plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java]: You can use BooleanUtils here.

File: plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java [hunk 2]
```diff
@@ -528,29 +533,35 @@ public void reserve(NicProfile nic, Network network, VirtualMachineProfile vm, D
...
-            if (staticNatIp != null) {
-                VlanVO staticNatVlan = _vlanDao.findById(staticNatIp.getVlanId());
-                vspStaticNat = _nuageVspEntityBuilder.buildVspStaticNat(null, staticNatIp, staticNatVlan, vspNic);
-            }
 
-            boolean defaultHasDns = getDefaultHasDns(networkHasDnsCache, nicFromDb);
-            VspDhcpVMOption dhcpOption = _nuageVspEntityBuilder.buildVmDhcpOption(nicFromDb, defaultHasDns, networkHasDns);
-            ReserveVmInterfaceVspCommand cmd = new ReserveVmInterfaceVspCommand(vspNetwork, vspVm, vspNic, vspStaticNat, dhcpOption);
-            Answer answer = _agentMgr.easySend(nuageVspHost.getId(), cmd);
+            if (!Boolean.TRUE.equals(vm.getParameter(VirtualMachineProfile.Param.RollingRestart))) {
```
Comment (rafaelweingartner) [inline plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java]: Would you mind extracting these blocks of code to specific methods and unit testing them? Some documentation might help as well. 

This is a comment to for this PR only: We really need to change our mind set and start using smaller, concise, well-documented and unit-tested methods. This in turn can greatly improve our productivity.

Comment (rafaelweingartner) [inline plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java]: You can use `BooleanUtils` to evaluate `vm.getParameter(VirtualMachineProfile.Param.RollingRestart))`

File: plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java [hunk 3]
```diff
@@ -723,6 +739,32 @@ public void deallocate(Network network, NicProfile nic, VirtualMachineProfile vm
             } else {
                 super.deallocate(network, nic, vm);
             }
+
+            if (virtualMachine.getType() == VirtualMachine.Type.DomainRouter) {
```
Comment (rafaelweingartner) [inline plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java]: Can you extract this blob to a method?

**Comment by rhtyd** [comment]
@fmaximus can you address outstanding review comments?

**Comment by rhtyd** [comment]
While one comment around refactoring code to a different method is outstanding, @fmaximus can you comment? I'm okay to merge this in order to unblock for RC2. /cc @PaulAngus @DaanHoogland

**Comment by fmaximus** [comment]
Still was running some tests on the refactoring.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
ok by me @rhtyd should it be packaged and smoke tested again? I saw you started that

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by fmaximus** [comment]
We found some issues with the fix, we're looking into it.

**Comment by rhtyd** [comment]
Test LGTM, let us know (soon) @fmaximus when you're good with your PR(s). Also, please use Github to log issues, thanks.

File: plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java (hunk 3 above)
Comment (DaanHoogland) [reply plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java]: I am pretty sure @rafaelweingartner means the entire if block and not the enum check, @fmaximus

**Comment by fmaximus** [comment]
It's good! :shipit:

**Comment by DaanHoogland** [comment]
@PaulAngus I did a walk through and am fine to merge this one, ok?

File: plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java (hunk 3 above)
Comment (rafaelweingartner) [reply plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java]: Spot on. I was referring to the IF body, and not to the IF condition. We really need shorter methods and more unit tests.

**Comment by rhtyd** [comment]
LGTM, based on code reviews and results I'm okay to merge this as well

**Pull request merged**